source/open_node/ast_visitor/ast_visitor.py

```diff
@@ -0,0 +1,74 @@
+from typing import Optional, TypeVar, Generic
+
+from open_node.ast import OpenNode
+
+T = TypeVar("T")
+
+
+class NodeVisitor(Generic[T]):
+    """
+    A node visitor base class that walks the abstract syntax tree and calls a
+    visitor function for every node found.  This function may return a value
+    which is forwarded by the `visit` method.
+
+    Modified from the implementation in ast.py in the Python standard library
+
+    The visit/generic_visit methods in subclasses do not have to have context argument.
+    """
+
+    def visit(self, node: OpenNode, context: Optional[T] = None):
+        """Visit a node."""
+        method = "visit_" + node.__class__.__name__
+        visitor = getattr(self, method, self.generic_visit)
+        # The visitor method may not have the context argument.
+        if context:
+            return visitor(node, context)
+        else:
+            return visitor(node)
+
+    def generic_visit(self, node: OpenNode, context: Optional[T] = None):
+        """Called if no explicit visitor function exists for a node."""
+        for _, value in node.__dict__.items():
+            if isinstance(value, list):
+                for item in value:
+                    if isinstance(item, OpenNode):
+                        if context:
+                            self.visit(item, context)
+                        else:
+                            self.visit(item)
+            elif isinstance(value, OpenNode):
+                if context:
+                    self.visit(value, context)
+                else:
+                    self.visit(value)
```

Comment:
Shorter to turn a single entry into a list before iterating:

```suggestion
            if not isinstance(value, list):
                value = [value]
            for item in value:
                if context:
                    self.visit(value, context)
                else:
                    self.visit(value)
```